### ISSUE TITLE
Unable to build latest version of pprof

### ISSUE THREAD

**Issue by ayushkarnawat** [NixOS/nixpkgs#138185]
### Describe the bug

TL;DR: Attempting to update the `pprof` pkg to support latest changes/bug fixes. Cannot build package as there is no cgo/SWIG to build c files when testing for symbolization on Mach-O binaries (see error below). However, when building using `go get -u github.com/google/pprof` in a nix-shell (see build instructions on main `google/pprof` [repo](https://github.com/google/pprof)), no such errors occur.

Here is the file to build latest `HEAD` of pprof within a `nix-shell` environment.
```nix
# shell.nix
# building pprof using go within nix-shell
with import <nixpkgs> {};

mkShell {
  name = "profiling";
  buildInputs = [
    gperftools
    go
    llvmPackages_11.llvm # llvm-symbolizer
    llvmPackages_11.clang # to link core libs
  ];
  # Automate process of extracting correct paths to binaries. This approach
  # works despite the fact that there might be multiple different versions of
  # the pkg of interest, as each version has its own unique hash.
  shellHook = ''
    clang_dir=$(ls -l /nix/store/ | grep ${llvmPackages_11.clang.pname + "-" +
      llvmPackages_11.clang.version} | grep '^d' | awk '/ / {print $NF}')
    export PATH="/nix/store/$clang_dir/bin:$PATH"

    # Use latest pprof (from github)
    # NOTE: delete go pkgs using `go get pprof@none`
    export GOPATH="$HOME/go/"
    go get -u github.com/google/pprof
    export PATH="$GOPATH/bin:$PATH"
  '';
}
```

#### Full details

Main goal/motivation: Profile code to identify areas which could benefit from optimization. To that end, we install gperftools (version 2.8), which includes a version of `pprof` to pretty print the results of the CPU profile. However, it does not properly symbolize the addresses to its respective function/line in the code. See example output below:

```shell
[user@system:~]$ pprof -text ./main main.prof
Using local file ./main.
Using local file main.prof.
Total: 871 samples
     351  40.3%  40.3%      351  40.3% 0x000000010ac0e24d
      70   8.0%  48.3%       70   8.0% 0x000000010ac0e248
      68   7.8%  56.1%       68   7.8% 0x000000010ac0e22c
      67   7.7%  63.8%       67   7.7% 0x000000010ac0e23e
      61   7.0%  70.8%       61   7.0% 0x000000010ac0e223
      58   6.7%  77.5%       58   6.7% 0x000000010ac0e280
      57   6.5%  84.0%       57   6.5% 0x000000010ac0e208
      53   6.1%  90.1%       53   6.1% 0x000000010ac0e213
      45   5.2%  95.3%       45   5.2% 0x000000010ac0e27b
      17   2.0%  97.2%       17   2.0% 0x000000010ac0e26a
       3   0.3%  97.6%        3   0.3% 0x000000010ac0e258
       2   0.2%  97.8%        2   0.2% 0x000000010ac0e226
       2   0.2%  98.0%        2   0.2% 0x000000010ac0e25b
       2   0.2%  98.3%        2   0.2% __ZNKSt3__16locale9use_facetERNS0_2idE
       1   0.1%  98.4%        1   0.1% 0x000000010ac0e210
       1   0.1%  98.5%        1   0.1% 0x000000010ac0e21c
       1   0.1%  98.6%        1   0.1% 0x000000010ac0e239
       1   0.1%  98.7%        1   0.1% 0x000000010ac0e240
       1   0.1%  98.9%        1   0.1% 0x000000010ac0e261
       1   0.1%  99.0%        1   0.1% 0x000000010ac0e268
       1   0.1%  99.1%        1   0.1% 0x000000010ac0e283
       1   0.1%  99.2%        1   0.1% 0x000000010ac0e28e
       1   0.1%  99.3%        1   0.1% 0x000000010ac0e293
       1   0.1%  99.4%        1   0.1% 0x000000010ac0e323
       1   0.1%  99.5%        1   0.1% 0x000000010ac0e333
       1   0.1%  99.7%        1   0.1% 0x000000010ac0e3f0
       1   0.1%  99.8%        1   0.1% 0x000000010ac0e40b
       1   0.1%  99.9%        1   0.1% 0x00007fff20548212
       1   0.1% 100.0%        1   0.1% 0x00007fff2054b2be
       0   0.0% 100.0%        1   0.1% 00007fff203a5726
       0   0.0% 100.0%        1   0.1% 00007fff203a6b7e
       0   0.0% 100.0%        1   0.1% 00007fff203aed8d
       0   0.0% 100.0%        1   0.1% 00007fff203befe4
       0   0.0% 100.0%        1   0.1% 00007fff2048b8f7
       0   0.0% 100.0%        1   0.1% 00007fff2048e24d
       0   0.0% 100.0%        1   0.1% 00007fff2048e526
       0   0.0% 100.0%      510  58.6% 00007fff20598f5c
       0   0.0% 100.0%        1   0.1% 0x000000010ac0e306
       0   0.0% 100.0%      505  58.0% 0x000000010ac0e31f
       0   0.0% 100.0%        1   0.1% 0x000000010ac0e344
       0   0.0% 100.0%        1   0.1% 0x000000010ac0e46c
       0   0.0% 100.0%      507  58.2% 0x000000010ac0e623
       0   0.0% 100.0%        2   0.2% 0x000000010ac0e6fc
       0   0.0% 100.0%        1   0.1% 0x000000010ac0e819
       0   0.0% 100.0%        1   0.1% 0x000000010ac0ed16
       0   0.0% 100.0%        1   0.1% 0x000000010ac0ed5a
       0   0.0% 100.0%        1   0.1% 0x000000010ac10860
       0   0.0% 100.0%        1   0.1% 0x000000010ac10a5c
       0   0.0% 100.0%        1   0.1% 0x000000010ac10b11
       0   0.0% 100.0%        1   0.1% 0x000000010ac10b98
       0   0.0% 100.0%        1   0.1% 0x000000011603c297
       0   0.0% 100.0%        2   0.2% 0x00007ffee4fee1a7
       0   0.0% 100.0%      479  55.0% 0x00007fff205c2d92
       0   0.0% 100.0%        1   0.1% __ZNSt3__111__stdoutbufIcE8overflowEi
       0   0.0% 100.0%        1   0.1% __ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEE3putEc
       0   0.0% 100.0%        2   0.2% __ZNSt3__113basic_ostreamIcNS_11char_traitsIcEEElsEj
       0   0.0% 100.0%        1   0.1% __Znwm
       0   0.0% 100.0%      510  58.6% __mh_execute_header
```

To fix this, the main developers of `pprof` recommend (see https://github.com/gperftools/gperftools/issues/791#issuecomment-233087038) to instead install the standalone version of `pprof` using `go`, which should be (relatively) straightforward as the package is available here. Normally, this is all we need to do: specify the standalone version of `pprof`, add that binary to the path (so that we are using the standalone version and not the version included with `gperftools`), and add either `binutils` or `llvm` for `addr2line` or `llvm-symbolizer`, respectively. Although this produces (human-readable) symbolizations, there are still some bugs/errors in the output (notice how this version cannot properly symbolize Mach-O binaries).

```shell
[user@system:~]$ PPROF_BINARY_PATH=./main
[user@system:~]$ pprof -text -trim=false main.prof
Local symbolization failed for libdyld.dylib: stat /usr/lib/system/libdyld.dylib: no such file or directory
Local symbolization failed for libsystem_platform.dylib: fat Mach-O archives are currently unsupported
Local symbolization failed for libsystem_kernel.dylib: fat Mach-O archives are currently unsupported
Local symbolization failed for libsystem_c.dylib: stat /usr/lib/system/libsystem_c.dylib: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libc++.1.dylib
Type: cpu
Showing nodes accounting for 8.09s, 100% of 8.09s total
      flat  flat%   sum%        cum   cum%
     8.02s 99.13% 99.13%      8.02s 99.13%  k_prime
     0.03s  0.37% 99.51%      8.02s 99.13%  primes
     0.03s  0.37% 99.88%      0.03s  0.37%  std::__1::locale::use_facet
     0.01s  0.12%   100%      0.01s  0.12%  [libsystem_kernel.dylib]
         0     0%   100%      8.06s 99.63%  <unknown>
         0     0%   100%      8.06s 99.63%  [libdyld.dylib]
         0     0%   100%      0.01s  0.12%  [libsystem_c.dylib]
         0     0%   100%      4.88s 60.32%  [libsystem_platform.dylib]
         0     0%   100%      8.06s 99.63%  main
         0     0%   100%      0.01s  0.12%  std::__1::__stdoutbuf::overflow
         0     0%   100%      0.04s  0.49%  std::__1::basic_ostream::operator<<
         0     0%   100%      0.01s  0.12%  std::__1::basic_ostream::put
         0     0%   100%      0.01s  0.12%  std::__1::endl
```

Given that recent versions of `pprof` address (some of) these bugs and add additional features, it is worthwhile to use a more updated version of `pprof` to allow for better profile reporting. Attempting to update to a recent version by changing the `rev` and associated `sha256` in the [default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/profiling/pprof/default.nix) file (similarly for the dependencies), and then build the file results in the following error below: It is unable to build the executable because it is not able to compile the `hello.c` or `lib.c` files without cgo/SWIG.

```shell
building '/nix/store/i8k11azazldg6xb1kfshbx4p4g44i7zf-pprof-unstable-2021-08-27.drv'...
unpacking sources
unpacking source archive /nix/store/ny13myl1rp35lca820k2c11jjpqy877b-pprof-02619b8
source root is pprof-02619b8
patching sources
configuring
unpacking source archive /nix/store/ngm69vin9qs4m4q291kzbi5fyxk0kr59-readline-2972be2
unpacking source archive /nix/store/sds05435lbg694gb4vd4r9qa85nf01fz-demangle-28f6c0f
building
internal/unsafeheader
internal/race
unicode/utf8
container/list
math/bits
runtime/internal/sys
internal/nettrace
unicode
sync/atomic
runtime/internal/atomic
runtime/cgo
internal/cpu
encoding
unicode/utf16
crypto/internal/subtle
runtime/internal/math
crypto/subtle
vendor/golang.org/x/crypto/cryptobyte/asn1
vendor/golang.org/x/crypto/internal/subtle
github.com/google/pprof/third_party/d3
github.com/google/pprof/third_party/d3flamegraph
github.com/google/pprof/third_party/svgpan
internal/bytealg
math
runtime
internal/reflectlite
sync
internal/singleflight
internal/testlog
math/rand
errors
sort
internal/oserror
path
io
vendor/golang.org/x/net/dns/dnsmessage
strconv
syscall
crypto/internal/randutil
text/tabwriter
hash
strings
bytes
hash/adler32
crypto/hmac
hash/crc32
crypto
reflect
crypto/rc4
vendor/golang.org/x/crypto/hkdf
vendor/golang.org/x/text/transform
bufio
html
regexp/syntax
internal/syscall/execenv
internal/syscall/unix
time
regexp
context
io/fs
internal/poll
internal/fmtsort
encoding/binary
crypto/x509/internal/macos
os
encoding/base64
crypto/sha1
vendor/golang.org/x/crypto/poly1305
crypto/cipher
crypto/sha256
crypto/sha512
crypto/md5
crypto/ed25519/internal/edwards25519
encoding/pem
vendor/golang.org/x/crypto/chacha20
crypto/des
crypto/aes
os/signal
vendor/golang.org/x/sys/cpu
vendor/golang.org/x/net/route
path/filepath
fmt
vendor/golang.org/x/crypto/chacha20poly1305
io/ioutil
os/exec
net
encoding/hex
flag
log
net/url
vendor/golang.org/x/crypto/curve25519
encoding/json
compress/flate
debug/dwarf
vendor/golang.org/x/text/unicode/norm
math/big
vendor/golang.org/x/net/http2/hpack
mime
mime/quotedprintable
net/http/internal
vendor/golang.org/x/text/unicode/bidi
github.com/ianlancetaylor/demangle
text/template/parse
compress/gzip
compress/zlib
vendor/golang.org/x/text/secure/bidirule
github.com/google/pprof/profile
vendor/golang.org/x/net/idna
text/template
debug/macho
debug/pe
debug/elf
github.com/google/pprof/internal/measurement
github.com/google/pprof/internal/graph
crypto/rand
crypto/dsa
encoding/asn1
crypto/elliptic
github.com/google/pprof/internal/elfexec
crypto/ed25519
crypto/rsa
html/template
crypto/x509/pkix
vendor/golang.org/x/crypto/cryptobyte
crypto/ecdsa
github.com/chzyer/readline
vendor/golang.org/x/net/http/httpproxy
net/textproto
crypto/x509
vendor/golang.org/x/net/http/httpguts
mime/multipart
crypto/tls
net/http/httptrace
net/http
github.com/google/pprof/internal/plugin
github.com/google/pprof/internal/symbolz
github.com/google/pprof/internal/transport
github.com/google/pprof/internal/binutils
github.com/google/pprof/internal/report
github.com/google/pprof/internal/symbolizer
github.com/google/pprof/internal/driver
github.com/google/pprof/driver
github.com/google/pprof
github.com/google/pprof/fuzz
package github.com/google/pprof/internal/binutils/testdata: C source files not allowed when not using cgo or SWIG: hello.c lib.c
builder for '/nix/store/i8k11azazldg6xb1kfshbx4p4g44i7zf-pprof-unstable-2021-08-27.drv' failed with exit code 1
error: build of '/nix/store/i8k11azazldg6xb1kfshbx4p4g44i7zf-pprof-unstable-2021-08-27.drv' failed
```

A short-term fix can be applied by updating to a [commit](https://github.com/google/pprof/commit/b9804c9f04c2d0c7bbd1af5129f853563f990541) which does not attempt to compile these files (as they are unavailable within the main pprof repo). From some initial tests, this compiles and generates plausible profiling reports. However, it means that we cannot update the package beyond this revision. This will be problematic in the future as new fixes/features are introduced; essentially we are side-stepping the main issue (see above for details).

A workaround to get the latest version of pprof working is to install `go` separately, set the GOPATH and install pprof (see build commands [here](https://github.com/google/pprof#building-pprof)), which should add that location of the binary in the path. Then, we can run `pprof` without any error and with all the latest features.

```shell
[user@system:~]$ PPROF_BINARY_PATH=./main
[user@system:~]$ pprof -text -trim=false main.prof
Local symbolization failed for libdyld.dylib: stat /usr/lib/system/libdyld.dylib: no such file or directory
Local symbolization failed for libsystem_c.dylib: stat /usr/lib/system/libsystem_c.dylib: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: libc++.1.dylib
Type: cpu
Showing nodes accounting for 8.19s, 100% of 8.19s total
      flat  flat%   sum%        cum   cum%
     8.13s 99.27% 99.27%      8.13s 99.27%  k_prime
     0.02s  0.24% 99.51%      0.02s  0.24%  [libsystem_kernel.dylib]
     0.02s  0.24% 99.76%      0.02s  0.24%  std::__1::__compressed_pair::first
     0.01s  0.12% 99.88%      7.68s 93.77%  primes
     0.01s  0.12%   100%      0.03s  0.37%  std::__1::__list_imp::__sz
         0     0%   100%      7.70s 94.02%  <unknown>
         0     0%   100%      7.70s 94.02%  [libdyld.dylib]
         0     0%   100%      0.02s  0.24%  [libsystem_c.dylib]
         0     0%   100%      0.11s  1.34%  [libsystem_platform.dylib]
         0     0%   100%      7.70s 94.02%  main
         0     0%   100%      0.02s  0.24%  std::__1::__stdoutbuf::overflow
         0     0%   100%      0.02s  0.24%  std::__1::basic_ostream::operator<<
         0     0%   100%      0.02s  0.24%  std::__1::basic_ostream::put
         0     0%   100%      0.02s  0.24%  std::__1::endl
         0     0%   100%      0.02s  0.24%  std::__1::list::size
```

Note that we can safely ignore the warnings that some of the `*.dylib` files cannot be symbolized as on MacOS Big Sur, Apple made changes such that dynamically linked libraries can no longer be searched as files within the system. Even if we were able to symbolize those libraries, this won't really matter as we will likely not be optimizing them anyways.

However, this approach is very cumbersome from a UX perspective, as it doesn't really build a derivation, but rather installs the binary in a user-defined path. As such, any changes would require the user to manually input commands to modify/delete the binaries. Instead, we should focus on making a working `pprof` derivation out of the box, such that all nix features work seamlessly.

So, to summarize, there are three approaches to solving this error:
1. Update revision to a "working", but old build (since it doesn't run test cases on c files)
2. Install go separately, build head commit of `pprof` (this works without error, albeit with a very cumbersome build UX)
3. Determine cause of why we are getting the SWIG/cgo errors when building in a nix-only environment (on builds that attempt run tests on C files). For future maintainability, this is the most robust approach to properly fix the issue (in my opinion).

### Steps To Reproduce
1. Update `pprof` package (and its dependencies) to latest revision(s):

```nix
# pprof/deps.nix
[
  {
    goPackagePath = "github.com/chzyer/readline";
    fetch = {
      type = "git";
      url = "https://github.com/chzyer/readline";
      rev = "2972be24d48e78746da79ba8e24e8b488c9880de";
      sha256 = "104q8dazj8yf6b089jjr82fy9h1g80zyyzvp3g8b44a7d8ngjj6r";
    };
  }
  {
    goPackagePath = "github.com/ianlancetaylor/demangle";
    fetch = {
      type = "git";
      url = "https://github.com/ianlancetaylor/demangle";
      rev = "28f6c0f3b63983aaa99575ca3b693afff7996387";
      sha256 = "0rsq3622gd40f1x1l7caidsxrmzg1993ich2higwd94fqbxs1r83";
    };
  }
]
```

```nix
# pprof/default.nix
{ lib, buildGoPackage, fetchgit }:

buildGoPackage rec {
  pname = "pprof-unstable";
  version = "2021-08-27";
  rev = "02619b876842e0d0afb5e5580d3a374dad740edb";

  goPackagePath = "github.com/google/pprof";

  src = fetchgit {
    inherit rev;
    url = "git://github.com/google/pprof";
    sha256 = "0j9y1a8hcbmd6rk0lmlqlxyzbnn9a0qnfh3g78cyip3wk9z47p0d";
  };

  goDeps = ./godeps.nix;

  meta = with lib; {
    description = "A tool for visualization and analysis of profiling data";
    homepage = "https://github.com/google/pprof";
    license = licenses.asl20;
    longDescription = ''
      pprof reads a collection of profiling samples in profile.proto format and
      generates reports to visualize and help analyze the data. It can generate
      both text and graphical reports (through the use of the dot visualization
      package).

      profile.proto is a protocol buffer that describes a set of callstacks and
      symbolization information. A common usage is to represent a set of sampled
      callstacks from statistical profiling. The format is described on the
      proto/profile.proto file. For details on protocol buffers, see
      https://developers.google.com/protocol-buffers

      Profiles can be read from a local file, or over http. Multiple profiles of
      the same type can be aggregated or compared.

      If the profile samples contain machine addresses, pprof can symbolize them
      through the use of the native binutils tools (addr2line and nm).

      This is not an official Google product.
    '';
  };
}
```

2. In a nix-shell, attempt to build the package

```nix
with import <nixpkgs> {};

mkShell {
  name = "profiling";
  buildInputs = [
    gperftools
    pprof
    llvmPackages_11.llvm # llvm-symbolizer
    llvmPackages_11.clang # to link core libs
  ];
  # Automate process of extracting correct paths to binaries. This approach
  # works despite the fact that there might be multiple different versions of
  # the pkg of interest, as each version has its own unique hash.
  shellHook = ''
    clang_dir=$(ls -l /nix/store/ | grep ${llvmPackages_11.clang.pname + "-" +
      llvmPackages_11.clang.version} | grep '^d' | awk '/ / {print $NF}')
    export PATH="/nix/store/$clang_dir/bin:$PATH"

    pprof_dir=$(ls -l /nix/store/ | grep ${pprof.pname + "-" + pprof.version} \
      | grep '^d' | awk '/ / {print $NF}')
    export PATH="/nix/store/$pprof_dir/bin:$PATH"
  '';
}
```

### Expected behavior

The `pprof` package builds properly and we are able to see correctly symbolized, human-readable profiling reports.

### Screenshots

Not necessary, but can attach if it would help.

### Additional context

N.A.

### Notify maintainers

Since you seem to have made major changes to these files:
@vdemeester 
@vcunat

### Metadata

```shell
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.5.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.8`
 - channels(ayushkarnawat): `"darwin-20.09pre246307.9b8530e3721, nixpkgs-21.03pre250093.0da76dab4c2"`
 - nixpkgs: `/Users/ayushkarnawat/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by vcunat** [comment]
I'm not good with darwin or go, so I most likely won't be of much help.  The last time I just kept using the "old" version of pprof from gperftools (on linux).

**Comment by ayushkarnawat** [comment]
Thanks @vcunat, I suspect that it has something to do with certain build instructions preventing cgo/SWIG from being detected when building the package. In fact, I tried compiling with both (and associated flags) to see if that would resolve the issue, but it didn't work (and yes, I checked that they were included in the PATH).

Digging deeper, I was curious as to how `buildGoPackage` is building packages, and thanks to documentation, found that it is [no longer the recommended](https://nixos.org/manual/nixpkgs/stable/#sec-language-go) method for building `go` packages. It is now deprecated in favor of `buildGoModule`. Rather than specifying the dependencies and associated `sha256` in a `deps.nix` file, the `buildGoModule` has an intermediate step which automatically fetches and builds an (intermediate) derivation of all the dependencies the main package requires, and uses that to build the final binary. Simply using this function resolved the issue.

I still do not understand why the original error is occurring when using `buildGoPackage`, but that probably does not matter now. I will open a PR with the fix (reproduced below for completeness):

```nix
# pprof/default.nix
{ lib, buildGoModule, fetchgit }:

buildGoModule rec {
  pname = "pprof-unstable";
  version = "2021-08-27";
  rev = "02619b876842e0d0afb5e5580d3a374dad740edb";

  src = fetchgit {
    inherit rev;
    url = "git://github.com/google/pprof";
    sha256 = "0j9y1a8hcbmd6rk0lmlqlxyzbnn9a0qnfh3g78cyip3wk9z47p0d";
  };

  # intermediate fetcher derivation
  vendorSha256 = "0yl8y3m2ia3cwxhmg1km8358a0225khimv6hcvras8r2glm69h3f";
  runVend = false;

  meta = with lib; {
    description = "A tool for visualization and analysis of profiling data";
    homepage = "https://github.com/google/pprof";
    license = licenses.asl20;
    longDescription = ''
      pprof reads a collection of profiling samples in profile.proto format
      and generates reports to visualize and help analyze the data. It can
      generate both text and graphical reports (through the use of the dot
      visualization package).

      profile.proto is a protocol buffer that describes a set of callstacks
      and symbolization information. A common usage is to represent a set of
      sampled callstacks from statistical profiling. The format is described
      on the proto/profile.proto file. For details on protocol buffers, see
      https://developers.google.com/protocol-buffers

      Profiles can be read from a local file, or over http. Multiple profiles
      of the same type can be aggregated or compared.

      If the profile samples contain machine addresses, pprof can symbolize
      them through the use of the native binutils tools (addr2line and nm).

      This is not an official Google product.
    '';
  };
}
```
